azure-iot-device/azure/iot/device/provisioning/transport/constant.py

```diff
@@ -0,0 +1,44 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+"""
+This module contains constants related to the transport package.
+"""
+
+__all__ = [
```

Comment:
Why are we only exposing a subset for import via "*"?

module specific `__all__` properties are relatively rare, is this really needed?

Reply:
Following up since this was unclear - when I made this comment in iteration 4, only a subset of items were being listed in the `__all__`, which confused me - why were only some of the constants being imported if someone did a `from constant import *`? That's what `__all__` does after all - it merely provides the list of things imported with `*`. Now, this is no longer an issue as all constants are included in the `__all__`.

But that then raises the second point - why is this even here? If everything in the file is included in the `__all__`... the `__all__` is entirely redundant. Because by default, `from constant import *` will import everything in the `constant` module that is not preceded by an convention-private underscore. This is why `__all__` is rare at module level - it's often quite redundant. It's only useful if there are things you want to explicitly include that do not have a convention-private underscore.

In fact, having it is a bit worse than redundant, because it is at high-risk for code rot. When you add a new constant, you now also need to remember to always add it to the `__all__` in this module, otherwise it won't be included with an `import *`.

The cap on all of this, is that `from <module> import *` is in general a terrible practice (with some exception, like how I've set up some of our `__init__.py` files), so for internal-only modules like this we shouldn't worry about what gets exposed with `*` since we just don't do that.

tl;dr there's no need for this `__all__` definition

---
In reply to: [282980495](https://github.com/Azure/azure-iot-sdk-python-preview/pull/81#discussion_r282980495) [](ancestors = 282980495)